File: packages/shared-types/src/graph.ts
import { DialogueNodeData, ConditionNodeData, ActionNodeData, NodeType } from './nodes';

export interface GraphCoordinates {
    x: number;
    y: number;
}

export type AnyNodeData = DialogueNodeData | ConditionNodeData | ActionNodeData;

export interface QuestNode<TData = AnyNodeData> {
    id: string;
    type: NodeType;
    position: GraphCoordinates;
    data: TData;
}

export interface QuestEdge {
    id: string;
    source: string;
    sourceHandle?: string;
    target: string;
    targetHandle?: string;
    animated?: boolean;
}

export interface QuestGraph {
    id: string;
    title: string;
    version?: number;
    rootNodeId?: string;
    nodes: QuestNode[];
    edges: QuestEdge[];
    metadata?: {
        createdAt?: string;
        updatedAt?: string;
        author?: string;
    };
}

export type GraphValidationErrorCode =
    | 'CYCLIC_DEPENDENCY'
    | 'UNREACHABLE_NODE'
    | 'DANGLING_EDGE'
    | 'EMPTY_ROOT'
    | 'EMPTY_GRAPH'
    | 'NO_ROOT_NODE'
    | 'DETECTED_CYCLE'
    | 'EMPTY_CHOICES';

export interface GraphValidationError {
    nodeId?: string;
    edgeId?: string;
    severity: 'error' | 'warning';
    code: GraphValidationErrorCode | string;
    message: string;
}

export interface GraphValidationResult {
    isValid: boolean;
    nodesCount?: number;
    edgesCount?: number;
    errors: GraphValidationError[];
}

export type TargetEngine = 'unity_csharp' | 'godot_gdscript' | 'json';

export interface ExportRequest {
    graph: QuestGraph;
    targetEngine: TargetEngine;
    namespace?: string;
}

export interface ExportResponse {
    fileName: string;
    targetEngine: TargetEngine;
    code: string;
}

export interface AIGenerateRequest {
    prompt: string;
    speakerName?: string;
}

export interface AIGenerateResponse {
    speaker: string;
    message: string;
    options: string[];
}

File: apps/api/app/schemas/graph.py
from typing import List, Dict, Any, Optional, Literal
from pydantic import BaseModel, Field

class NodePosition(BaseModel):
    x: float
    y: float

class QuestNodeSchema(BaseModel):
    id: str
    type: str
    position: Optional[NodePosition] = None
    data: Dict[str, Any]

class QuestEdgeSchema(BaseModel):
    id: str
    source: str
    target: str
    sourceHandle: Optional[str] = None
    targetHandle: Optional[str] = None

class QuestGraphPayload(BaseModel):
    id: str
    title: str = "Untitled Quest"
    rootNodeId: Optional[str] = None
    nodes: List[QuestNodeSchema]
    edges: List[QuestEdgeSchema]

class ValidationErrorItem(BaseModel):
    nodeId: Optional[str] = None
    edgeId: Optional[str] = None
    severity: Literal['error', 'warning', 'info']
    code: str
    message: str

class ValidationResponse(BaseModel):
    isValid: bool
    nodesCount: int
    edgesCount: int
    errors: List[ValidationErrorItem]

class ExportRequest(BaseModel):
    graph: QuestGraphPayload
    targetEngine: Literal['unity_csharp', 'godot_gdscript', 'json_compact']
    namespace: Optional[str] = "QuestCraft.Generated"

class ExportResponse(BaseModel):
    fileName: str
    targetEngine: str
    code: str
